core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -3104,23 +3104,106 @@ public static int position(ByteString seek, ByteString s) {
 
   /** SQL {@code POSITION(seek IN string FROM integer)} function. */
   public static int position(String seek, String s, int from) {
-    final int from0 = from - 1; // 0-based
-    if (from0 > s.length() || from0 < 0) {
+    if (from == 0) {
+      throw RESOURCE.fromNotZero().ex();
+    }
+    // Case when from is positive
+    if (from > 0) {
+      final int from0 = from - 1; // 0-based
+      if (from0 >= s.length() || from0 < 0) {
+        return 0;
+      } else {
+        return s.indexOf(seek, from0) + 1;
+      }
+    }
+    // Case when from is negative
+    final int rightIndex = from + s.length(); // negative position to positive index
+    if (rightIndex <= 0) {
       return 0;
     }
-
-    return s.indexOf(seek, from0) + 1;
+    return s.substring(0, rightIndex + 1).lastIndexOf(seek) + 1;
   }
 
   /** SQL {@code POSITION(seek IN string FROM integer)} function for byte
    * strings. */
   public static int position(ByteString seek, ByteString s, int from) {
-    final int from0 = from - 1;
-    if (from0 > s.length() || from0 < 0) {
+    if (from == 0) {
+      throw RESOURCE.fromNotZero().ex();
+    }
+    // Case when from is positive
+    if (from > 0) {
+      final int from0 = from - 1; // 0-based
+      if (from0 >= s.length() || from0 < 0) {
+        return 0;
+      } else {
+        return s.indexOf(seek, from0) + 1;
+      }
+    }
+    // Case when from is negative
+    final int rightIndex = from + s.length();
+    if (rightIndex <= 0) {
       return 0;
     }
+    int lastIndex = 0;
+    while (lastIndex < rightIndex) {
+      int indexOf = s.substring(lastIndex, rightIndex + 1).indexOf(seek) + 1;
+      if (indexOf == 0) {
+        break;
+      }
+      lastIndex += indexOf;
+    }
+    return lastIndex;
+  }
 
-    return s.indexOf(seek, from0) + 1;
+  /** SQL {@code POSITION(seek, string, from, occurrence)} function. */
+  public static int position(String seek, String s, int from, int occurrence) {
+    if (occurrence == 0) {
+      throw RESOURCE.occurrenceNotZero().ex();
+    }
+    for (int i = 0; i < occurrence; i++) {
```

Comment:
`from` is constant and `i` is not. So would it make sense to move the `if` outside the `for`, so that it's only checked once?

Reply:
Not sure I understand what you mean. From gets updated every time the for loop runs, so we have to check whether to use the logic for from being positive or negative each time. I think checking if from is 0 after assignment is necessary because it will error if we continue running

Reply:
Ok, `from` isn't constant. But is `from > 0` constant? Or, if it isn't, should it be? That is, if we go into the `else` branch of the `if` on i = 0, we also want to go into the `else` branch for i = 1 and i = 2.